File: src/java/magma/option/Some.java
package magma.option;

import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

public record Some<T>(T value) implements Option<T> {
    @Override
    public <R> Option<R> map(Function<T, R> mapper) {
        return new Some<>(mapper.apply(value));
    }

    @Override
    public T orElseGet(Supplier<T> other) {
        return value;
    }

    @Override
    public Tuple<Boolean, T> toTuple(T other) {
        return new Tuple<>(true, value);
    }

    @Override
    public void ifPresent(Consumer<T> consumer) {
        consumer.accept(value);
    }

    @Override
    public T orElse(T other) {
        return value;
    }

    @Override
    public Option<T> filter(Predicate<T> predicate) {
        return predicate.test(value) ? this : new None<>();
    }

    @Override
    public boolean isPresent() {
        return true;
    }
}

File: src/clang/magma/option/Some.c
#include "Some.h"
magma.option.Option<magma.option.R> map(magma.option.R(*mapper)(T)){return (mapper.apply(value));
}
T orElseGet(T(*other)()){return value;
}
magma.option.Tuple<int, T> toTuple(T other){return (true, value);
}
magma.option.void ifPresent(void(*consumer)(T)){consumer.accept(value);
}
T orElse(T other){return value;
}
magma.option.Option<T> filter(int(*predicate)(T)){return predicate.test(value)?this:();
}
int isPresent(){return true;
}
magma.option.R match(magma.option.R(*ifPresent)(T), magma.option.R(*ifEmpty)()){return ifPresent.apply(value);
}
int isEmpty(){return false;
}
magma.option.Option<T> or(magma.option.Option<T>(*other)()){return this;
}
magma.option.Option<magma.option.R> flatMap(magma.option.Option<magma.option.R>(*mapper)(T)){return mapper.apply(value);
}

